### PR TITLE
SONARJS-746 Support new export syntax

### PR DESCRIPTION
**Pull request by vilchik-elena**

### CODE REVIEW

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+public interface ExportDefaultBindingWithExportList extends Tree {
+
+  IdentifierTree exportedDefaultIdentifier();
+
+  SyntaxToken commaToken();
+
+  SpecifierListTree exportList();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 2]
```diff
@@ -0,0 +1,39 @@
...
+
+  FromClauseTree fromClause();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 3]
```diff
@@ -0,0 +1,39 @@
...
+
+  @Nullable
+  SyntaxToken semicolonToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.api.tree.declaration;
+
+import javax.annotation.Nullable;
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+public interface ExportDefaultBindingWithExportList extends Tree {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public interface. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 5]
```diff
@@ -0,0 +1,39 @@
...
+
+  IdentifierTree exportedDefaultIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+
+  SyntaxToken commaToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java [hunk 7]
```diff
@@ -194,11 +197,22 @@ public void visitNameSpaceExportDeclaration(NameSpaceExportDeclarationTree tree)
...
-
   public void visitNamedExportDeclaration(NamedExportDeclarationTree tree) {
     scanChildren(tree);
   }
 
+  public void visitExportDefaultBinding(ExportDefaultBinding tree) {
+    scanChildren(tree);
+  }
+
+  public void visitExportDefaultBindingWithNameSpaceExport(ExportDefaultBindingWithNameSpaceExport tree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java [hunk 8]
```diff
@@ -194,11 +197,22 @@ public void visitNameSpaceExportDeclaration(NameSpaceExportDeclarationTree tree)
...
+    scanChildren(tree);
+  }
+
+
+  public void visitExportDefaultBindingWithExportList(ExportDefaultBindingWithExportList tree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java [hunk 9]
```diff
@@ -194,11 +197,22 @@ public void visitNameSpaceExportDeclaration(NameSpaceExportDeclarationTree tree)
     scanChildren(tree);
   }
 
-
   public void visitNamedExportDeclaration(NamedExportDeclarationTree tree) {
     scanChildren(tree);
   }
 
+  public void visitExportDefaultBinding(ExportDefaultBinding tree) {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/visitors/DoubleDispatchVisitor.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+public interface ExportDefaultBindingWithNameSpaceExport extends Tree {
+
+  IdentifierTree exportedDefaultIdentifier();
+
+  SyntaxToken commaToken();
+
+  SyntaxToken starToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 11]
```diff
@@ -0,0 +1,43 @@
...
+
+  SyntaxToken asToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 12]
```diff
@@ -0,0 +1,43 @@
...
+
+  IdentifierTree synonymIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 13]
```diff
@@ -0,0 +1,43 @@
...
+
+  FromClauseTree fromClause();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+
+  @Nullable
+  SyntaxToken semicolonToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 15]
```diff
@@ -0,0 +1,43 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.api.tree.declaration;
+
+import javax.annotation.Nullable;
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+public interface ExportDefaultBindingWithNameSpaceExport extends Tree {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public interface. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 16]
```diff
@@ -0,0 +1,43 @@
...
+
+  IdentifierTree exportedDefaultIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 17]
```diff
@@ -0,0 +1,43 @@
...
+
+  SyntaxToken commaToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/NameSpaceExportDeclarationTree.java [hunk 18]
```diff
@@ -35,6 +36,12 @@
 
   SyntaxToken starToken();
 
+  @Nullable
+  SyntaxToken asToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/NameSpaceExportDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/NameSpaceExportDeclarationTree.java [hunk 19]
```diff
@@ -35,6 +36,12 @@
...
+
+  @Nullable
+  IdentifierTree synonymIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/NameSpaceExportDeclarationTree.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java [hunk 20]
```diff
@@ -0,0 +1,35 @@
...
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+public interface ExportDefaultBinding extends Tree {
+
+  IdentifierTree exportedDefaultIdentifier();
+
+  FromClauseTree fromClause();
+
+  @Nullable
+  SyntaxToken semicolonToken();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java [hunk 21]
```diff
@@ -0,0 +1,35 @@
...
+ * Inc., 51 Franklin Street, Fifth Floor, Boston, MA  02110-1301, USA.
+ */
+package org.sonar.plugins.javascript.api.tree.declaration;
+
+import javax.annotation.Nullable;
+import org.sonar.plugins.javascript.api.tree.Tree;
+import org.sonar.plugins.javascript.api.tree.expression.IdentifierTree;
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+public interface ExportDefaultBinding extends Tree {
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java]: MINOR Document this public interface. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java [hunk 22]
```diff
@@ -0,0 +1,35 @@
...
+
+  IdentifierTree exportedDefaultIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java [hunk 23]
```diff
@@ -0,0 +1,35 @@
...
+
+  FromClauseTree fromClause();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by SonarTech** [comment]
SonarQube analysis reported 23 issues
* MINOR 23 minor

Watch the comments in this conversation to review them.

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java [hunk 24]
```diff
@@ -0,0 +1,44 @@
...
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+/**
+ * This interface stands for the tail of export declaration (proposed for ES2017)
+ * <pre>export A, {B as BB, C} from 'moduleName';</pre>
+ * <pre>export {@link #exportedDefaultIdentifier()}, {@link #exportList()} {@link #fromClause()} ;</pre>
+ */
+public interface ExportDefaultBindingWithExportList extends Tree {
+
+  IdentifierTree exportedDefaultIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithExportList.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java [hunk 25]
```diff
@@ -0,0 +1,48 @@
...
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+/**
+ * This interface stands for the tail of export declaration (proposed for ES2017)
+ * <pre>export A, * as B from 'moduleName';</pre>
+ * <pre>export {@link #exportedDefaultIdentifier()}, * as {@link #synonymIdentifier()} {@link #fromClause()} ;</pre>
+ */
+public interface ExportDefaultBindingWithNameSpaceExport extends Tree {
+
+  IdentifierTree exportedDefaultIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBindingWithNameSpaceExport.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

File: javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java [hunk 26]
```diff
@@ -0,0 +1,40 @@
...
+import org.sonar.plugins.javascript.api.tree.lexical.SyntaxToken;
+
+/**
+ * This interface stands for the tail of export declaration (proposed for ES2017)
+ * <pre>export A from 'moduleName';</pre>
+ * <pre>export {@link #exportedDefaultIdentifier()} {@link #fromClause()};</pre>
+ */
+public interface ExportDefaultBinding extends Tree {
+
+  IdentifierTree exportedDefaultIdentifier();
```
Comment (SonarTech) [inline javascript-frontend/src/main/java/org/sonar/plugins/javascript/api/tree/declaration/ExportDefaultBinding.java]: MINOR Document this public method. [rule](https://dory.sonarsource.com/sonarqube/coding_rules#rule_key=squid%3AUndocumentedApi)

**Comment by SonarTech** [comment]
SonarQube analysis reported 20 issues
* MINOR 20 minor

Watch the comments in this conversation to review them.

**Comment by ivandalbosco** [comment]
Seems very good. I had a hard time matching the contents of JavaScriptGrammar#NAMED_EXPORT_DECLARATION with the 7 cases defined in the spec in http://leebyron.com/ecmascript-export-default-from, section 1.1.2, but your tests are exhaustive.

How is supposed to resolve the conflicts: the developer or the reviewer?

**Comment by vilchik-elena** [comment]
@ivandalbosco I will resolve them. Thanks!

**Comment by vilchik-elena** [comment]
@ivandalbosco done

**Pull request merged**